Migrate NPC rules to a new, dedicated imgui menu

RenechCDDA opened the pull request:
#### Summary
Interface "NPC follower rules can now be viewed and modified in a dedicated, mouse-supported window (imgui)"

#### Purpose of change
Everyone agrees the dialogue menu does its job for this, but it could be so much better...

#### Describe the solution
Make it better, and with making it better comes mouse support!

There is now a dedicated top-level talk topic which opens this new imgui window, no need to remember which of the sub-menus your specific rule was in. Each rule can be fully adjusted with a dedicated hotkey or mouse-sensitive buttons. For rules with multiple states (e.g. CBM recharge %) keyboard presses *of the hotkey* will cycle to the next setting, while mouse presses are set directly to the selected % button. Keyboard users also have the option of navigating(standard up/down/left/right inputs) directly to the desired setting and press enter, instantly setting that value without a need to toggle through the proceeding ones.

~~(Still WIP) Tie the imgui buttons to a specific set of auto-assigned hotkeys, like our dialogue does. This is almost certainly a requirement before this PR can be undrafted.~~

#### Describe alternatives you've considered


#### Testing
https://github.com/CleverRaven/Cataclysm-DDA/assets/84619419/86334d68-ef69-4324-b7ce-54275e9b94ca




#### Additional context
In the interests of encouraging reviews, I have added a hopefully funny easter egg somewhere in this PR. It will only appear in game with a 0.1% chance.

RenechCDDA committed:
Migrate NPC rules to a new, dedicated imgui menu

Kamejeir commented:
clicking the x on the menu makes it unresponsive until re-opened (also no indication to hit esc or space for closing it, in general)

RenechCDDA commented:
> clicking the x on the menu makes it unresponsive until re-opened (also no indication to hit esc or space for closing it, in general)

I am aware, thank you.

RenechCDDA commented:
"Hm I don't know how to solve this, how did I do it on the mission UI?"

> the mission UI doesn't close when you press the close button either

THANKS ME

RenechCDDA committed:
Follower rules UI: Hotkeys,focus follows selection, can close window by pressing close button.

Also fixed this in mission ui

RenechCDDA commented:
I'm marking this ready for review, because that's what it is: Ready for review.

I am reasonably confident in it, though I may still push more commits. There was a reasonable request to switching from using "CHANGE" buttons to checkboxes. There's also some boilerplate which could stand to be templated.

RenechCDDA committed:
Make rules with multi-selections into a template

RenechCDDA commented:
> I'm marking this ready for review, because that's what it is: Ready for review.
> 
> I am reasonably confident in it, though I may still push more commits. There was a reasonable request to switching from using "CHANGE" buttons to checkboxes. There's also some boilerplate which could stand to be templated.

Instead of checkboxes, we got colored buttons. (Checkboxes could still be added later, but colorizing the buttons keeps it consistent and means I don't have to do anything else!)

Template was implemented to keep nice and streamlined.

The only thing I could want now is a confirmation on defaulting all settings. I was using query_yn for this, but it causes a crash(see code comments).

Assuming tests pass, we are good to go.

akrieger commented on src/npctalk_rules.cpp:
```diff
@@ -0,0 +1,309 @@
+#include "ui_manager.h"
+#include "imgui/imgui.h"
+
+static std::map<cbm_recharge_rule, std::string> recharge_map = {
+    {cbm_recharge_rule::CBM_RECHARGE_ALL, "<ally_rule_cbm_recharge_all_text>" },
+    {cbm_recharge_rule::CBM_RECHARGE_MOST, "<ally_rule_cbm_recharge_most_text>" },
+    {cbm_recharge_rule::CBM_RECHARGE_SOME, "<ally_rule_cbm_recharge_some_text>" },
+    {cbm_recharge_rule::CBM_RECHARGE_LITTLE, "<ally_rule_cbm_recharge_little_text>" },
+    {cbm_recharge_rule::CBM_RECHARGE_NONE, "<ally_rule_cbm_recharge_none_text>" },
+};
+
+static std::map<cbm_reserve_rule, std::string> reserve_map = {
+    {cbm_reserve_rule::CBM_RESERVE_ALL, "<ally_rule_cbm_reserve_all_text>" },
+    {cbm_reserve_rule::CBM_RESERVE_MOST, "<ally_rule_cbm_reserve_most_text>" },
+    {cbm_reserve_rule::CBM_RESERVE_SOME, "<ally_rule_cbm_reserve_some_text>" },
+    {cbm_reserve_rule::CBM_RESERVE_LITTLE, "<ally_rule_cbm_reserve_little_text>" },
+    {cbm_reserve_rule::CBM_RESERVE_NONE, "<ally_rule_cbm_reserve_none_text>" },
+};
+
+static std::map<combat_engagement, std::string> engagement_rules = {
+    {combat_engagement::NONE, "<ally_rule_engagement_none>" },
+    {combat_engagement::CLOSE, "<ally_rule_engagement_close>" },
+    {combat_engagement::WEAK, "<ally_rule_engagement_weak>" },
+    {combat_engagement::HIT, "<ally_rule_engagement_hit>" },
+    {combat_engagement::ALL, "<ally_rule_engagement_all>" },
+    {combat_engagement::FREE_FIRE, "<ally_rule_engagement_free_fire>" },
+    {combat_engagement::NO_MOVE, "<ally_rule_engagement_no_move>" },
+};
+
+static std::map<aim_rule, std::string> aim_rule_map = {
+    {aim_rule::WHEN_CONVENIENT, "<ally_rule_aim_when_convenient>" },
+    {aim_rule::SPRAY, "<ally_rule_aim_spray>" },
+    {aim_rule::PRECISE, "<ally_rule_aim_precise>" },
+    {aim_rule::STRICTLY_PRECISE, "<ally_rule_aim_strictly_precise>" },
+};
```
Unless these need to be internally sorted or the key isn't hashable, it's generally preferred to use `unordered_map`

RenechCDDA replied:
Right, I misunderstood you and tried to apply this. The ordering is in fact important, as it's used for the hotkey wrapping. std::map::upper_bound lets us press the hotkey and move through the map

akrieger commented on src/npctalk_rules.cpp:
```diff
@@ -0,0 +1,309 @@
+    }
+}
+
+template<typename T>
+void follower_rules_ui_impl::multi_rule_header( std::string id, T &rule,
+        std::map<T, std::string> rule_map, bool should_advance )
```
```suggestion
template<typename T>
void follower_rules_ui_impl::multi_rule_header( const std::string &id, T &rule,
        const std::map<T, std::string> &rule_map, bool should_advance )
```

akrieger commented on src/npctalk_rules.h:
```diff
@@ -0,0 +1,61 @@
+        // Prepares for a rule option with multiple valid selections. Advances and wraps through
+        // those options as the hotkey is pressed.
+        template<typename T>
+        void multi_rule_header( std::string id, T &rule, std::map<T, std::string> rule_map,
```
```suggestion
        void multi_rule_header( const std::string &id, T &rule, const std::map<T, std::string> &rule_map,
```

RenechCDDA committed:
...Move settings around between your followers. Why not

RenechCDDA committed:
final(?) adjustments

db48x commented on src/npctalk_rules.cpp:
```diff
@@ -0,0 +1,518 @@
+        ImGui::PopID();
+    }
+
+    // Engagement rules require their own set of buttons, each instruction is unique
+    ImGui::Separator();
+    ImGui::NewLine();
+    assigned_hotkey = input_ptr->next_unassigned_hotkey( hotkeys, assigned_hotkey );
+    print_hotkey( assigned_hotkey );
+    multi_rule_header( "ENGAGEMENT_RULES", guy->rules.engagement, engagement_rules,
+                       pressed_key == assigned_hotkey );
+    int engagement_rule_number = 0;
+    for( const std::pair<const combat_engagement, std::string> &engagement_rule : engagement_rules ) {
+        ImGui::PushID( rule_number );
+        rule_number++;
+        engagement_rule_number++;
+        // Could use a better label for these...
+        std::string button_label = std::to_string( engagement_rule_number );
+        ImGui::SameLine();
+        if( guy->rules.engagement == engagement_rule.first ) {
+            ImGui::PushStyleColor( ImGuiCol_Button, c_green );
+        }
+        if( ImGui::Button( button_label.c_str() ) ) {
+            guy->rules.engagement = engagement_rule.first;
+            ImGui::SetKeyboardFocusHere( -1 );
+        }
+        ImGui::PopStyleColor();
+        ImGui::PopID();
+    }
+    ImGui::SameLine();
+    draw_colored_text( _( "Engagement rules" ), c_white );
+    ImGui::NewLine();
+    draw_colored_text( string_format( "%s", get_parsed( engagement_rules[guy->rules.engagement] ) ) );
```
This business of looping over a list of rules is rather repetitive. Pull it out into a function? Call it `rule_list` or something? I see that for `ally_rules` the body is somewhat different though, so maybe some thought is required.

db48x commented on src/mission_ui.cpp:
```diff
@@ -104,7 +104,7 @@ void mission_ui::draw_mission_ui()
         p_impl.last_action = ctxt.handle_input();
 
-        if( p_impl.last_action == "QUIT" ) {
+        if( p_impl.last_action == "QUIT" || !p_impl.get_is_open() ) {
```
Make this a separate bug fix please.

RenechCDDA replied:
I agree with you in principle but it should be wholly uncontroversial and I've just learned

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/84619419/2b9886f2-03d4-4181-89ce-827be607a19a)

...So I'm going to leave it in.

db48x reviewed (changes requested):
I think that the boolean options should be checkboxes, and the multiple–choice ones should be list boxes or combo boxes.

db48x commented on src/npctalk_rules.cpp:
```diff
@@ -0,0 +1,518 @@
+            guy->rules.toggle_flag( this_rule.rule );
+            guy->rules.toggle_specific_override_state( this_rule.rule, !rule_enabled );
+        }
+        ImGui::PopStyleColor();
```
This pop needs to be guarded by the same condition as the push, so that they pair up correctly.

db48x reviewed (changes requested):
And the same for all the other repetitions of the same pattern in the rest of the code.

db48x commented:
Another thing, the import/export dialog is really confusing if you only have one NPC, because the table is empty.

RenechCDDA commented on src/npctalk_rules.cpp:
```diff
@@ -0,0 +1,518 @@
+        return;
+    }
+
+    ImGui::InvisibleButton( "TOP_OF_WINDOW_KB_SCROLL_SELECTABLE", ImVec2() );
```
This and the bottom of the window one are for scrolling purposes. Otherwise you could scroll all the way down and your nav ends up... above the last line of text, so you cannot see what the last rule is without specifically switching to pageup/pagedown or moving the scrollbar with the mouse.

RenechCDDA replied:
To illustrate: 

https://github.com/CleverRaven/Cataclysm-DDA/assets/84619419/28569666-b5ce-4d9b-b1d4-f3d56e3597a2

RenechCDDA commented:
> Another thing, the import/export dialog is really confusing if you only have one NPC, because the table is empty.

...Is it? It ends up being an empty table, which makes sense to me.

RenechCDDA committed:
Hopefully truly last touches to the follower rules UI
-Extract the common button setup into a function
-Only pop style color when it's been pushed
-Export/Import displays warning when only 1 follower
-other very minor changes

db48x commented:
Normally one exports things to a file, so it made no sense until I went back and read the source code. Maybe the wording should be changed, and perhaps there should be some acknowledgement that you haven’t got any other NPCs to make it more clear that you should return to this window later.

db48x commented on src/npctalk_rules.cpp:
```diff
@@ -0,0 +1,456 @@
+        } else {
+            rules_text = string_format( "%s", get_parsed( this_rule.rule_false_text ) );
+        }
+        bool was_clicked = false;
+        std::string label = _( "Toggle" );
+        setup_button( rule_number, label, rule_enabled, was_clicked );
+        if( was_clicked || pressed_key == assigned_hotkey ) {
```
It would be much simpler for `setup_button` (and `setup_table_button`) to simply return true or false.

RenechCDDA commented:
> Normally one exports things to a file, so it made no sense until I went back and read the source code. Maybe the wording should be changed, and perhaps there should be some acknowledgement that you haven’t got any other NPCs to make it more clear that you should return to this window later.

OK, done
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/84619419/63ed6c37-188c-45f9-83c4-87fae1c1264b)

RenechCDDA committed:
Numpad and hjkl support for follower rules UI

db48x commented on src/npctalk_rules.cpp:
```diff
@@ -0,0 +1,467 @@
+    } else if( last_action == "RIGHT" ) {
+        ImGui::NavMoveRequestSubmit( ImGuiDir_Right, ImGuiDir_Right, ImGuiNavMoveFlags_None,
+                                     ImGuiScrollFlags_None );
+    } else if( last_action == "CONFIRM" ) {
+        ImGui::ActivateItem( ImGui::GetFocusID() );
+    }
```
Oops, I just noticed this bug. The Button handles the enter key itself, so this code double–activates the button. When you press enter the setting will flicker as it is toggled on and then immediately back off. You can just remove this clause.

RenechCDDA replied:
It does not handle numpad enter.

db48x replied:
Awkward.

RenechCDDA replied:
Marking as resolved, then.

db48x commented:
Checkboxes and radio buttons were not that hard to do:

![Screenshot from 2024-06-12 14-42-47](https://github.com/CleverRaven/Cataclysm-DDA/assets/228849/2d39ace5-79ff-450e-936f-98144e5f2ea5)

Want me to send you a pull request?

Brambor commented:
Not for this issue to solve, but:
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/13402666/a77436d0-7c33-4a4e-8db0-201ea75b0f7a)

RenechCDDA commented:
> Want me to send you a pull request?

I'm not enthused about the idea. I don't think it'd be wrong, I just... prefer? the colored buttons. Basically,

-I like the colored buttons. (100% subjective opinion)

-The colored buttons offers a more CDDA-style look (again, 100% subjective opinion)

-I do not think checkboxes would convey what is happening in the import/export screen, which means regardless the button implementation stays there.
--If the buttons stay there, it clashes with this part of the rules screen which would subsequently use checkboxes.

Points in favor of checkboxes:
-Checkboxes offer better accessibility, guaranteeing that any flavor of colorblindness or visual impairment can use them.

-Your checkbox implementation looks pretty sexy

-If you already have the branch, any cost of switching to it has already been paid

So if it's not any trouble to you, please send the PR . Or PR directly to cleverraven if/after this one is merged, if that would be easier for you.

> Not for this issue to solve, but:

It previously used <mypronoun> which resulted in `will attacking anything She wants.` (notice the capitalization)

I was trying to keep the strings as close to the old ones as possible, but might be worth kicking that one to something different.

db48x commented:
Did you by any chance change your github preferences so that you don’t get pull requests in your fork? It doesn’t give me the option of sending you one.

You could also just pull from my branch and push it to yours, if that’s easier than mucking about with github.

RenechCDDA commented:
I don't believe I've made any changes, and just recently got the first and only PR to my fork.

But sure, you can just link the branch. I can merge, cherry pick, etc as appropriate

db48x commented:
My branch is https://github.com/db48x/Cataclysm-DDA/tree/follower_rules_UI

db48x commented:
BTW, I did manage to make a PR the other day: https://github.com/RenechCDDA/Cataclysm-DDA/pull/2

IoIqq commented:
This looks really cool!

Maleclypse commented:
There’s conflicts to resolve if this is ready for merge?

RenechCDDA commented:
I need to rebase it (to keep the commit history clean-ish) and then check out db48x's branch, I just haven't gotten around to it despite constantly expecting to

db48x commented:
I’m not busy today so I’ll go ahead and rebase for you. I’ll end up with a new PR though.

RenechCDDA commented:
> I’m not busy today so I’ll go ahead and rebase for you. I’ll end up with a new PR though.

Well I guess that WOULD let me work on something else...
